kostal/__init__.py

```diff
@@ -18,88 +18,182 @@
+            self._keytable = self._key2td
+            self.datastucture = "html"
+        else:
+            self._keytable = self._key2json
+            self.datastucture = "json"
 
     def run(self):
+        """
+        Run method for the plugin
+        """
+        self.logger.debug("run method Kostal called")
         self.alive = True
         self._sh.scheduler.add('Kostal', self._refresh, cycle=self.cycle)
 
     def stop(self):
+        """
+        Stop method for the plugin
+        """
+        self.logger.debug("stop method Kostal called")
         self.alive = False
 
     def parse_item(self, item):
-        if 'kostal' in item.conf:
-            kostal_key = item.conf['kostal']
-            if kostal_key in self._key2td:
-                self._items.append([item, kostal_key])
-                return self.update_item
-            else:
-                logger.warn('invalid key {0} configured', kostal_key)
-        return None
-
+        """
+        Default plugin parse_item method. Is called when the plugin is initialized.
+        Selects each item corresponding to its attribute keywords and adds it to an internal array
+        :param item: The item to process.
+        """
+        if self.has_iattr(item.conf, 'kostal'):
+            self._items[self.get_iattr_value(item.conf, 'kostal')] = item
+            self.logger.debug("parse item: {0}".format(item))
+            return self.update_item
     def parse_logic(self, logic):
         pass
-
     def update_item(self, item, caller=None, source=None, dest=None):
-        if caller != 'Kostal':
-            pass
-
+        if item():
+            if self.has_iattr(item.conf, 'kostal'):
+                self.logger.debug("update_item ws called with item '{}' from caller '{}', source '{}' and dest '{}'".format(item, caller, source, dest))
+                pass
     def _refresh(self):
         start = time.time()
-        try:
-            data = self._sh.tools.fetch_url(
-                'http://' + self.ip + '/', self.user, self.passwd, timeout=2).decode()
-            # remove all attributes for easy findall()
-            data = re.sub(r'<([a-zA-Z0-9]+)(\s+[^>]*)>', r'<\1>', data)
-            # search all TD elements
-            table = re.findall(r'<td>([^<>]*)</td>', data, re.M | re.I | re.S)
-            for kostal_key in self._key2td:
-                value = table[self._key2td[kostal_key]].strip()
-                if 'x x x' not in value:
-                    logger.debug('set {0} = {1}'.format(kostal_key, value))
-                    self._values[kostal_key] = value
-            for item_cfg in self._items:
-                if item_cfg[1] in self._values:
-                    item_cfg[0](self._values[item_cfg[1]], 'Kostal')
-        except Exception as e:
-            logger.error(
-                'could not retrieve data from {0}: {1}'.format(self.ip, e))
-            return
-
+        if self.datastucture == "html":
+            #HTML-OLD-Coding
```

Comment:
I would move the code for the old and new datastructure handling into separate methods. Additionally I would assign this method to the `datastructure` attribute instead of a simple string (see init method) and simply invoke it `self.datastructure()` (this will the invoke the assigned method from init method).

This would also shrink the method _refresh().